### PR TITLE
feat: implement partial_params support to McpToolSpec

### PR DESCRIPTION
**Pull request by martimfasantos**
## Overview

This PR adds support for the `partial_params` feature in `McpToolSpec`, enabling developers to pre-fill tool parameters that should be set deterministically rather than being exposed to LLMs or agents.

It now supports **both global parameters** (applied to all tools) and **tool-specific parameters**, allowing flexible parameter injection and overrides.

## Motivation

When integrating with internal APIs through MCP, there are often scenarios where certain parameters should not be exposed to the LLM or client:

### Common Use Cases:

1. **Authentication & Authorization**

   * API keys, tokens, or user IDs injected server-side
   * Session identifiers determined by context
   * Tenant IDs in multi-tenant systems

2. **Infrastructure Context**

   * Environment identifiers (prod/staging/dev)
   * Region or datacenter parameters
   * Service endpoints that vary by deployment

This allows sensitive information to remain hidden while simplifying agent logic and reducing token usage.

## Example

**Before:** LLM needs to know and provide parameters like `user_id` and `api_key`:

```python
spec = McpToolSpec(client=mcp_client)
# Tool schema exposes: {"query": str, "user_id": str, "api_key": str}
```

**After:** Parameters are pre-filled and hidden from the LLM.

### Global Parameters (applied to all tools)

```python
McpToolSpec(client, global_partial_params={"api_key": "xxx"})
```

### Tool-Specific Parameters

```python
McpToolSpec(client, partial_params_by_tool={
    "search_tool": {"api_key": "xxx"},
    "weather_tool": {"location": "NYC"}
})
```

### Combined Global + Tool-Specific Overrides

```python
McpToolSpec(
    client,
    global_partial_params={"api_key": "default"},
    partial_params_by_tool={
        "search_tool": {"query": "preset"},    # Adds on top of global
        "weather_tool": {"api_key": None}      # Removes global param
    }
)
```

**Behavior:**

* Global params are always applied first
* Tool-specific params merge with or override global params
* Setting a value to `None` removes that global param for the specific tool

## Implementation Details

* **Schema Filtering**: `partial_params` removes pre-filled fields from the tool's input schema before presenting it to the LLM
* **Automatic Injection**: Parameters are automatically injected during tool execution
* **Pydantic v2 Compatible**: Properly handles Pydantic v2 model field annotations
* **Multiple Tools**: Applies consistently across all tools fetched from the MCP server

### Core Changes

* Added `global_partial_params` and `partial_params_by_tool` to `McpToolSpec.__init__`
* Modified `to_tool_list_async()` to filter schema fields based on `partial_params`
* Added `remove_model_fields()` utility
* Passes `partial_params` to `FunctionTool.from_defaults()` for execution-time injection

### Testing

Added comprehensive test coverage (9 new tests):

* Field removal from tool schemas (sync and async)
* Multiple field removal
* Pydantic model compatibility
* Edge cases (empty dict, non-existent fields, all fields removed)
* Optional field handling
* Multi-tool scenarios

## Benefits

* **Enhanced Security**: Prevents sensitive parameters from appearing in LLM prompts or logs
* **Simplified Agent Logic**: LLMs only see parameters they need to reason about
* **Cleaner Integrations**: Reduces token usage and cognitive load on the model
* **Flexible Architecture**: Supports dynamic parameter injection based on runtime context
* **Backwards Compatible**: Optional parameters; existing code continues to work unchanged

## Testing

All tests pass (26/26):

```bash
pytest llama-index-integrations/tools/llama-index-tools-mcp/tests/test_mcp.py -v
```

## Migration Guide

No breaking changes. To use this feature, simply add `global_partial_params` and/or `partial_params_by_tool` when initializing `McpToolSpec`:

```python
spec = McpToolSpec(
    client=mcp_client,
    allowed_tools=["search_database"],
    global_partial_params={"api_key": "xxx"},
    partial_params_by_tool={
        "search_tool": {"user_id": get_current_user_id()},
        "weather_tool": {"location": "NYC"}
    }
)
```

---

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Your pull-request will likely not be merged unless it is covered by some form of impactful unit testing.

- [x] I added new unit tests to cover this change
- [ ] I believe this change is already covered by existing unit tests

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `uv run make format; uv run make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/tools/llama-index-tools-mcp/llama_index/tools/mcp/base.py [hunk 1]
```diff
@@ -133,13 +135,19 @@ async def to_tool_list_async(self) -> List[FunctionTool]:
             model_schema = self.create_model_from_json_schema(
                 tool.inputSchema, model_name=f"{tool.name}_Schema"
             )
+            if self.partial_params:
+                model_schema = self.remove_model_fields(
+                    model_schema,
+                    fields_to_remove=set(self.partial_params),
```
Comment (logan-markewich) [inline llama-index-integrations/tools/llama-index-tools-mcp/llama_index/tools/mcp/base.py]: an MCP server can have multiple tools. I guess this assumes that partial params applies to all tools?

Might be nice to optional pass a dict of `tool-name -> partial_params` or a master list of params to apply to all tools?

Comment (martimfasantos) [reply llama-index-integrations/tools/llama-index-tools-mcp/llama_index/tools/mcp/base.py]: Yes it does - your suggestion makes a lot of sense! Thanks

I just added support for both use cases:

**`global_partial_params`**: Apply params to all tools
```python
McpToolSpec(client, global_partial_params={"api_key": "xxx"})
```

**`partial_params_by_tool`**: Per-tool customization
```python
McpToolSpec(client, partial_params_by_tool={
    "search_tool": {"api_key": "xxx"},
    "weather_tool": {"location": "NYC"}
})
```

**Combined**: Global params + tool-specific overrides
```python
McpToolSpec(
    client,
    global_partial_params={"api_key": "default"},
    partial_params_by_tool={
        "search_tool": {"query": "preset"},    # Adds on top of global
        "weather_tool": {"api_key": None}     # Removes global param
    }
)
```

- Global params are always applied first
- Tool-specific params merge with/override global
- Set value to `None` to remove a global param for a specific tool
